src/Microsoft.Health.Fhir.Core/Features/Operations/Import/ImportOrchestratorTask.cs

```diff
@@ -0,0 +1,507 @@
+            }
+            catch (ImportFileEtagNotMatchException eTagEx)
+            {
+                _logger.LogInformation(eTagEx, "Import file etag not match.");
+
+                errorResult = new ImportTaskErrorResult();
+                errorResult.HttpStatusCode = HttpStatusCode.BadRequest;
+                errorResult.ErrorMessage = eTagEx.Message;
+
+                taskResultData = new TaskResultData(TaskResult.Fail, JsonConvert.SerializeObject(errorResult));
+            }
+            catch (ImportProcessingException processingEx)
+            {
+                _logger.LogInformation(processingEx, "Failed to process input resources.");
+
+                errorResult = new ImportTaskErrorResult();
+                errorResult.HttpStatusCode = HttpStatusCode.BadRequest;
+                errorResult.ErrorMessage = processingEx.Message;
+
+                taskResultData = new TaskResultData(TaskResult.Fail, JsonConvert.SerializeObject(errorResult));
+            }
+            catch (Exception ex)
+            {
+                _logger.LogInformation(ex, "Failed to import data.");
+
+                errorResult = new ImportTaskErrorResult();
+                errorResult.HttpStatusCode = HttpStatusCode.InternalServerError;
+                errorResult.ErrorMessage = ex.Message;
+
+                throw new RetriableTaskException(JsonConvert.SerializeObject(errorResult));
+            }
+
+            if (_orchestratorTaskContext.Progress > ImportOrchestratorTaskProgress.InputResourcesValidated)
+            {
+                // Post-process operation cannot be cancelled.
+                try
+                {
+                    await _fhirDataBulkImportOperation.DeleteDuplicatedResourcesAsync(CancellationToken.None);
+                    await _fhirDataBulkImportOperation.PostprocessAsync(CancellationToken.None);
+
+                    _logger.LogInformation("Postprocess Completed");
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogInformation(ex, "Failed at postprocess step.");
+
+                    ImportTaskErrorResult postProcessEerrorResult = new ImportTaskErrorResult();
+                    postProcessEerrorResult.HttpStatusCode = HttpStatusCode.InternalServerError;
+                    postProcessEerrorResult.ErrorMessage = ex.Message;
+                    postProcessEerrorResult.InnerError = errorResult; // other error if any.
+
+                    throw new RetriableTaskException(JsonConvert.SerializeObject(postProcessEerrorResult));
+                }
+            }
+
+            if (taskResultData == null) // No exception
+            {
+                taskResultData = new TaskResultData(TaskResult.Success, JsonConvert.SerializeObject(_orchestratorTaskContext.ImportResult));
+            }
+
+            return taskResultData;
+        }
+
+        public void Cancel()
+        {
+            if (!_cancellationTokenSource.IsCancellationRequested)
```

Comment:
Is it safer to check if it is null? _cancellationTokenSource

Reply:
Yes, we need to check.